File: depth_first.py
# iterative way of solving problem
def depthFirst(arr, start):
    stack = [start]
    
    while(len(stack) > 0):
        current = stack.pop();
        print(current)
        
        for i in range(len(arr[current])):
            stack.append(arr[current][i])

# recursive way of solving problem
def depthFirstRecursive(arr,start):
    print(start)
    for i in range(len(arr[start])):
        depthFirstRecursive(arr, arr[start][i])
    
graph = {
    'a': ['b', 'c'],
    'b': ['d'],
    'c': ['e'],
    'd': ['f'],
    'e': [],
    'f': [],
}

depthFirst(graph, 'a')
depthFirstRecursive(graph, 'a')

File: depth_first.php
<?php

// iterative way of solving problem
function depthFirst($arr, $start) {
    $stack = [$start];
    
    while(count($stack) > 0) {
        $current = array_pop($stack);
        echo $current;
        
        foreach ($arr[$current] as $neighbour) {
            array_push($stack, $neighbour);
        }
    }
}

// recursive way of solving problem
function depthFirstRecursive($arr, $start) {
    echo $start;
    foreach($arr[$start] as $neighbour) {
        depthFirstRecursive($arr, $neighbour);
    }
}

$graph = [
    'a' => ['b', 'c'],
    'b' => ['d'],
    'c' => ['e'],
    'd' => ['f'],
    'e' => [],
    'f' => [],
];

depthFirst($graph, 'a');
echo "<br>";
depthFirstRecursive($graph, 'a');
?>
